matrix_sdk/src/client.rs

```diff
@@ -534,6 +533,83 @@ impl Client {
         })
     }
 
+    /// Creates a new client for making HTTP requests to the homeserver of the
+    /// given user. Follows homeserver discovery directions described
+    /// [here](https://spec.matrix.org/unstable/client-server-api/#well-known-uri).
+    ///
+    /// # Arguments
+    ///
+    /// * `user_id` - The id of the user whose homeserver the client should
+    ///   connect to.
+    ///
+    /// # Example
+    /// ```no_run
+    /// # use std::convert::TryFrom;
+    /// # use matrix_sdk::{Client, identifiers::UserId};
+    /// # use futures::executor::block_on;
+    /// let alice = UserId::try_from("@alice:example.org").unwrap();
+    /// # block_on(async {
+    /// let client = Client::new_from_user_id(alice.clone()).await.unwrap();
+    /// client.login(alice.localpart(), "password", None, None).await.unwrap();
+    /// # });
+    /// ```
+    pub async fn new_from_user_id(user_id: UserId) -> Result<Self> {
+        let config = ClientConfig::new();
+        Client::new_from_user_id_with_config(user_id, config).await
+    }
+
+    /// Creates a new client for making HTTP requests to the homeserver of the
+    /// given user and configuration. Follows homeserver discovery directions
+    /// described [here](https://spec.matrix.org/unstable/client-server-api/#well-known-uri).
+    ///
+    /// # Arguments
+    ///
+    /// * `user_id` - The id of the user whose homeserver the client should
+    ///   connect to.
+    ///
+    /// * `config` - Configuration for the client.
+    pub async fn new_from_user_id_with_config(
+        user_id: UserId,
+        config: ClientConfig,
+    ) -> Result<Self> {
+        let homeserver = Client::homeserver_from_user_id(user_id)?;
+        let mut client = Client::new_with_config(homeserver, config)?;
+
+        let well_known = client.discover_homeserver().await?;
+        let well_known = Url::parse(well_known.homeserver.base_url.as_ref())?;
+        client.set_homeserver(well_known);
+        client.get_supported_versions().await?;
+        Ok(client)
+    }
+
+    fn homeserver_from_user_id(user_id: UserId) -> Result<Url> {
+        let homeserver = format!("https://{}", user_id.server_name());
+        Ok(Url::parse(homeserver.as_str())?)
+    }
+
+    async fn discover_homeserver(&self) -> Result<discover_homeserver::Response> {
+        self.send(discover_homeserver::Request::new(), Some(RequestConfig::new().disable_retry()))
+            .await
+    }
+
+    fn set_homeserver(&mut self, homeserver_url: Url) {
+        self.homeserver = Arc::new(homeserver_url);
```

Comment:
I think we should put `self.homeserver` behind a `RwLock` to get rid of the mutable borrow here. Handling well-known in the `login` response will require this as well.

Reply:
Done, although this made a couple functions `async` by necessity and `Debug` for `Client` got much less useful.